### PR TITLE
[nl server] get default and enabled indexes from deployment

### PR DESCRIPTION
**Pull request by chejennifer**
Get default index and enabled indexes from deployment because this could be instance specific

### CODE REVIEW

File: deploy/helm_charts/dc_website/values.yaml [hunk 1]
```diff
@@ -136,7 +136,11 @@ nl:
   models:
   memory: "2G"
   workers: 1
-  vertex_ai_models:
+  embeddingsSpec:
+    defaultIndex: ""
+    enabledIndexes: []
```
Comment (shifucun) [inline deploy/helm_charts/dc_website/values.yaml:141]: given defaultIndex, shall we make this "additionalIndexes"?

File: deploy/helm_charts/dc_website/values.yaml [hunk 2]
```diff
@@ -136,7 +136,11 @@ nl:
...
+    vertexAIModels:
+    rankingEnabled: false
```
Comment (shifucun) [inline deploy/helm_charts/dc_website/values.yaml:143]: enableReranking

File: deploy/helm_charts/envs/autopush.yaml [hunk 3]
```diff
@@ -44,27 +44,31 @@ serviceAccount:
...
-      project_id: datcom-website-dev
-      location: us-central1
-      prediction_endpoint_id: "3977846152316846080"
-    cross-encoder-mxbai-rerank-base-v1:
-      project_id: datcom-website-dev
-      location: us-central1
-      prediction_endpoint_id: "284894457873039360"
+  embeddingsSpec:
+    defaultIndex: "medium_ft"
+    enabledIndexes: ["medium_ft", "sdg_ft", "undata_ft", "bio_ft", "medium_lance_ft", "medium_vertex_ft", "base_uae_mem", "medium_vertex_mistral"]
```
Comment (shifucun) [inline deploy/helm_charts/envs/autopush.yaml:49]: nit: one line for per item and sorted?

File: nl_server/loader.py [hunk 4]
```diff
@@ -166,3 +186,31 @@ def get_env_path(flask_env: str, file_name: str) -> str:
...
+    embeddings_spec_dict = _CUSTOM_DC_EMBEDDINGS_SPEC
+  # otherwise try to get from gke.
+  elif os.path.exists(_EMBEDDINGS_SPEC_PATH):
+    with open(_EMBEDDINGS_SPEC_PATH) as f:
+      embeddings_spec_dict = json.load(f) or {}
+  # If that path doesn't exist, assume we are running locally and use the values
+  # from autopush.
+  else:
+    current_file_path = Path(__file__)
+    autopush_env_values = f'{current_file_path.parent.parent}/deploy/helm_charts/envs/autopush.yaml'
```
Comment (shifucun) [inline nl_server/loader.py:207]: nit: do not use autopush in the variable names to make this adaptive.

File: nl_server/loader.py [hunk 5]
```diff
@@ -107,15 +127,15 @@ def _load_yaml(flask_env: str) -> Dict[str, any]:
...
+def _update_app_config(app: Flask, attribute_model: NLAttributeModel,
                        nl_embeddings: emb_map.EmbeddingsMap,
                        embeddings_version_map: Dict[str, any],
-                       vertex_ai_models: Dict[str, Dict] = None):
+                       embeddings_spec: EmbeddingsSpec):
   app.config[config.ATTRIBUTE_MODEL_KEY] = attribute_model
   app.config[config.NL_EMBEDDINGS_KEY] = nl_embeddings
   app.config[config.NL_EMBEDDINGS_VERSION_KEY] = embeddings_version_map
-  app.config[config.VERTEX_AI_MODELS_KEY] = vertex_ai_models or {}
+  app.config[config.DEFAULT_INDEX_TYPE_KEY] = embeddings_spec.default_index
```
Comment (pradh) [inline nl_server/loader.py:137]: nit: should we just get the default index from app.config[EMBEDDINGS_SPEC_KEY]?

File: nl_server/loader.py [hunk 6]
```diff
@@ -23,12 +26,30 @@
...
+@dataclass
+class EmbeddingsSpec:
+  default_index: str
+  enabled_indexes: List[str]
+  vertex_ai_model_info: Dict[str, any]
+  ranking_enabled: bool
+
+
+# TODO: move this to be part of custom dc deployment
+_CUSTOM_DC_EMBEDDINGS_SPEC = {
```
Comment (pradh) [inline nl_server/loader.py:48]: I thought you had a separate file before, that's better?

File: nl_server/loader.py [hunk 7]
```diff
@@ -23,12 +26,30 @@
...
+    'defaultIndex': 'medium_ft',
+    'enabledIndexes': ['medium_ft'],
```
Comment (pradh) [inline nl_server/loader.py:50]: should this include `custom_ft` ?

File: nl_server/loader.py (hunk 4 above)
Comment (pradh) [inline nl_server/loader.py:207]: nit: use constant?

File: nl_server/loader.py [hunk 8]
```diff
@@ -166,3 +186,31 @@ def get_env_path(flask_env: str, file_name: str) -> str:
...
+    current_file_path = Path(__file__)
+    autopush_env_values = f'{current_file_path.parent.parent}/deploy/helm_charts/envs/autopush.yaml'
+    with open(autopush_env_values) as f:
+      autopush_env = yaml.full_load(f)
+      embeddings_spec_dict = autopush_env['nl']['embeddingsSpec']
+
+  return EmbeddingsSpec(
+      embeddings_spec_dict['defaultIndex'],
+      embeddings_spec_dict['enabledIndexes'],
+      vertex_ai_model_info=embeddings_spec_dict.get('vertexAIModels') or {},
```
Comment (pradh) [inline nl_server/loader.py:215]: nit: why not dict.get('vertexAIModels', {})

File: nl_server/loader.py [hunk 9]
```diff
@@ -166,3 +186,31 @@ def get_env_path(flask_env: str, file_name: str) -> str:
...
+  else:
+    current_file_path = Path(__file__)
+    autopush_env_values = f'{current_file_path.parent.parent}/deploy/helm_charts/envs/autopush.yaml'
+    with open(autopush_env_values) as f:
+      autopush_env = yaml.full_load(f)
+      embeddings_spec_dict = autopush_env['nl']['embeddingsSpec']
+
+  return EmbeddingsSpec(
+      embeddings_spec_dict['defaultIndex'],
+      embeddings_spec_dict['enabledIndexes'],
```
Comment (pradh) [inline nl_server/loader.py:214]: be tolerant to missing field .get('enabledIndexes', []) ?

File: nl_server/loader.py [hunk 10]
```diff
@@ -166,3 +186,31 @@ def get_env_path(flask_env: str, file_name: str) -> str:
...
+      vertex_ai_model_info=embeddings_spec_dict.get('vertexAIModels') or {},
+      ranking_enabled=embeddings_spec_dict['rankingEnabled'])
```
Comment (pradh) [inline nl_server/loader.py:216]: nit: .get('rankingEnabled', False) .. be tolerant to missing field

File: nl_server/loader.py [hunk 11]
```diff
@@ -23,12 +26,30 @@
...
 _CUSTOM_EMBEDDINGS_YAML_PATH = 'datacommons/nl/custom_embeddings.yaml'
+_EMBEDDINGS_SPEC_PATH: str = '/datacommons/nl/embeddings_spec.json'
+
+
+@dataclass
+class EmbeddingsSpec:
+  default_index: str
+  enabled_indexes: List[str]
+  vertex_ai_model_info: Dict[str, any]
+  ranking_enabled: bool
```
Comment (pradh) [inline nl_server/loader.py:44]: nit: should this be in config.py along with other configs?

File: nl_server/config.py [hunk 12]
```diff
@@ -107,80 +98,31 @@ class EmbeddingsConfig:
...
-          prediction_endpoint_id=vertex_ai_model_info[model_name]
-          ['prediction_endpoint_id'],
-          location=vertex_ai_model_info[model_name]['location'])
-    else:
-      raise AssertionError(
-          'Error parsing information for model {model_name}: unsupported type {model_type}'
-      )
+def parse_v1(embeddings_map: Dict[str, any], vertex_ai_model_info: Dict[str,
+                                                                        any],
+             ranking_enabled: bool) -> EmbeddingsConfig:
```
Comment (pradh) [inline nl_server/config.py:118]: nit: why not pass EmbeddingsSpec

File: nl_server/config.py [hunk 13]
```diff
@@ -205,11 +147,34 @@ def parse_v1(embeddings_map: Dict[str, any],
...
 
-  return EmbeddingsConfig(indexes=indexes, models=models)
-
+  # parse the models
+  models = {}
+  for model_name, model_info in embeddings_map.get('models', {}).items():
+    # Only parse the model if it is used by an index or it is a ranking model
+    # and ranking is allowed. Otherwise, skip
+    if model_name not in used_models:
+      if not ranking_enabled or model_info['usage'] != ModelUsage.RERANKING:
```
Comment (pradh) [inline nl_server/config.py:156]: nit: use a _is_model_enabled(model_name) helper?

**Review by pradh — COMMENTED** [review]
Feel free to ignore the nits for now.

File: nl_server/custom_dc_constants.py [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Temporary constants used for custom dc that should be read from deployment.
+TODO: move these constants to be available in deployment"""
+
+CUSTOM_DC_VERTEX_AI_MODELS = {}
+CUSTOM_DC_EMBEDDING_INDEX_INFO = {
+    'default': 'medium_ft',
+    'enabled': ['medium_ft']
+}
```
[the remark this answers is not in the document]
Comment (pradh) [reply nl_server/custom_dc_constants.py]: Stale comment from before, no longer valid.

File: deploy/helm_charts/dc_website/values.yaml (hunk 1 above)
Comment (chejennifer) [reply deploy/helm_charts/dc_website/values.yaml:141]: as discussed offline, leaving as is

File: deploy/helm_charts/dc_website/values.yaml (hunk 2 above)
Comment (chejennifer) [reply deploy/helm_charts/dc_website/values.yaml]: done

File: deploy/helm_charts/envs/autopush.yaml (hunk 3 above)
Comment (chejennifer) [reply deploy/helm_charts/envs/autopush.yaml]: done

File: nl_server/loader.py (hunk 4 above)
Comment (chejennifer) [reply nl_server/loader.py]: done

File: nl_server/custom_dc_constants.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (chejennifer) [reply nl_server/custom_dc_constants.py:22]: this has been updated

File: nl_server/loader.py (hunk 5 above)
Comment (chejennifer) [reply nl_server/loader.py]: done

File: nl_server/loader.py (hunk 6 above)
Comment (chejennifer) [reply nl_server/loader.py]: done

File: nl_server/loader.py (hunk 7 above)
Comment (chejennifer) [reply nl_server/loader.py]: I'm not sure if custom dc can have index names other than custom_ft, so I made it where these are the enabled indexes from the big catalog & then just add all additional custom indexes

File: nl_server/loader.py (hunk 4 above)
Comment (chejennifer) [reply nl_server/loader.py]: done

File: nl_server/loader.py (hunk 8 above)
Comment (chejennifer) [reply nl_server/loader.py:206]: when coming from the environment, the key will exist but value will be None and in both cases where key doesn't exist OR value is none, i want to use {}

File: nl_server/loader.py (hunk 9 above)
Comment (chejennifer) [reply nl_server/loader.py]: done

File: nl_server/loader.py (hunk 10 above)
Comment (chejennifer) [reply nl_server/loader.py]: done

File: nl_server/loader.py (hunk 11 above)
Comment (chejennifer) [reply nl_server/loader.py]: i kept this here because it's only used in the loader, can move if you prefer

File: nl_server/config.py (hunk 12 above)
Comment (chejennifer) [reply nl_server/config.py]: EmbeddingsSpec has things not used so felt more clear to pass the things we actully need, wdyt

File: nl_server/config.py (hunk 13 above)
Comment (chejennifer) [reply nl_server/config.py]: done

File: nl_server/loader.py (hunk 8 above)
Comment (pradh) [reply nl_server/loader.py:206]: comment please :)

File: nl_server/loader.py (hunk 7 above)
Comment (pradh) [reply nl_server/loader.py]: I'm not sure, something to check.  Lets get this in and test on custom dc

File: nl_server/loader.py (hunk 11 above)
Comment (pradh) [reply nl_server/loader.py]: ok, sg

File: nl_server/config.py (hunk 12 above)
Comment (pradh) [reply nl_server/config.py]: sg for now.

**Review by shifucun — APPROVED** [review]

File: nl_server/routes.py [hunk 15]
```diff
@@ -35,7 +35,8 @@
 
 @bp.route('/healthz')
 def healthz():
-  default_index_type = current_app.config[config.DEFAULT_INDEX_TYPE_KEY]
+  default_index_type = current_app.config[
+      config.EMBEDDINGS_SPEC_KEY]['default_index']
```
Comment (pradh) [inline nl_server/routes.py:39]: Don't crash (bug from yesterday).  How about, if 'default_index' is empty, return Service Unavailable?

Comment (pradh) [inline nl_server/routes.py:39]: is it `defaultIndex`?  Use constants?

Comment (chejennifer) [reply nl_server/routes.py]: done

Comment (chejennifer) [reply nl_server/routes.py]: no, EmbeddingsSpec object uses default_index. updated to constants

Comment (pradh) [reply nl_server/routes.py]: Hmm, if its a class, why aren't we using `.default_index` ?

**Review by pradh — APPROVED** [review]

File: nl_server/routes.py (hunk 15 above)
Comment (chejennifer) [reply nl_server/routes.py]: you're right, updated

**Comment by chejennifer** [comment]
thanks for the reviews!

**Pull request merged**